django/utils/text.py

```diff
@@ -103,38 +110,46 @@ def handle_endtag(self, tag):
             self.tags.remove(tag)
 
     def handle_data(self, data):
-        if self.func == "words":
-            data = re.split(r"(?<=\S)\s(?=\S)", data)
-            output = escape(" ".join(data[: self.num]))
-        else:
-            output = escape("".join(data[: self.num]))
-
+        data, output = self.process(data)
         data_len = len(data)
-        if self.num < data_len:
-            if self.func == "words" and self.truncate and self.truncate[0] == " ":
-                output = output.strip()
+        if self.remaining < data_len:
+            self.remaining = 0
             self.output += Truncator.add_truncation_text(output, self.truncate)
-            self.output += "".join([f"</{tag}>" for tag in self.tags])
-            raise self.TruncationCompleted()
-        self.num -= data_len
+            raise self.TruncationCompleted
+        self.remaining -= data_len
         self.output += output
 
-    def _truncate_html(self, func, text, num, truncate):
```

Comment:
This felt a bit messy. Instead we use a more standard convention in the call sites:

```python
parser = CustomHTMLParser(...)
parser.feed(data)
parser.close()
return parser.output  # Extract our parsed/manipulated data.
```